ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcSplit.java

```diff
@@ -384,7 +384,6 @@ public void parse(Configuration conf, Path rootPath) throws IOException {
         OrcRawRecordMerger.TransactionMetaData.findWriteIDForSynthetcRowIDs(getPath(), rootPath, conf);
     writeId = tmd.syntheticWriteId;
     stmtId = tmd.statementId;
-    bucketId = AcidUtils.parseBucketId(getPath());
```

Comment:
Is it safe to remove this? Do we have test coverage around this method (directly or indirectly)?

Reply:
Previously bucketId was set in parse() method. Now bucketId is set in constructor. Both assign the same value.

Reply:
Yes but do we *always* pass through the constructor which sets the `bucketId`?

The `*Split` classes are serializable which makes me a bit skeptical about which constructor is called at each point.   

Moreover, `writeId` and `stmtId` are also set here and they seem to depend exclusively on `getPath()` which again makes me a bit suspicious why they were not set in constructor if was possible to do so.

Reply:
Hey @ayushtkn are you familiar with `InputSplit` and `InputFormat`? Do you know if the refactoring here is safe?

Reply:
As I understand we always pass through constructor which sets bucketId. The class has 2 constructors and the bucket Id is set in both of them. Otherwise it wouldn't compile, I made bucketId final and final members must be set before the end of a constructor. 

writeId and stmtId depend also on configuration and rootPath which are passed as arguments to parse method, but bucket Id was depending only on path.
https://github.com/apache/hive/pull/3606/files#diff-f33c72f4823ff07b5fb662959932648b9e96c52f248e44df29bd4e8d550487b2R383-R387:~:text=TransactionMetaData%20tmd%20%3D-,OrcRawRecordMerger,-.TransactionMetaData.

Reply:
I am not very sure about this part of code. But just going through the code. the ``bucketId= AcidUtils.parseBucketId(getPath());``, the Path can change also if someone calls ``readFields``. It does this in the super class:
```
  public void readFields(DataInput in) throws IOException {
    file = new Path(Text.readString(in));
```
the `file` is what returns from ``getPath()``, So, I am not sure if it is a good idea to just have that set in just the constructor.

Moreover if any other change in Hadoop comes, which bothers the path in some other method post the construction invocation, again we will be prone to may be some silent errors.

Second, there is a legacy protected constructor which doesn't have path and expect ``readFields`` to do the magic for it, if some thirdparty application/project is still using that by any means, that will get screwed, So, we have to remove that constructor also I feel.

Would require someone more experienced to tell, how safe this is

Reply:
If we mark a `transient` field as `final` we are basically breaking the class since we can no longer set the value of the `bucketId` field during deserialization.

It may be fine to set `bucketId` in the constructor but I am pretty sure we cannot remove this line from the `parse` method.

Reply:
Thanks for the code review! I didn't think that this class may be serizlized/deserialized and bucketId may need to be set after deserialization in the parse method. Will close this PR.

Reply:
Really sorry for making you invest time on this; I didn't foresee all this when I made the initial comment.

Reply:
No problem at all! This way I learned something new.